File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        #currMax = [0]
        res = 0
        R = len(grid)
        C = len(grid[0])
        for r in range(R):
            for c in range(C):
                if grid[r][c] == 1:
                    res = max(self.dfs(grid, r, c), res)
        return res

    def dfs(self, grid: List[List[int]], r: int, c: int) -> int:
        # if node in visited: return 
        # add to visited set [1, 1]
        # explore neighbors:
        #   dfs()
        if r < 0 or c < 0 or r >= len(grid) or c >= len(grid[0]) or grid[r][c] == 0:
            return 0
        grid[r][c] = 0 # mark as visited
        dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]]
        count = 1
        for di in dirs:
            nr = r + di[0]
            nc = c + di[1]
            count += self.dfs(grid, nr, nc)
        
        return count
        

File: Data Structures & Algorithms/clone-graph/submission-1.java
/*
Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if(node == null) return null;
        Map<Node, Node> visited = new HashMap<>();
        Node res = dfs(node, visited);
        return res;
    }

    public Node dfs(Node node, Map<Node, Node> visited) {
        if(visited.containsKey(node)) return visited.get(node);
        
        //add to visited set
        Node clone = new Node(node.val);
        visited.put(node, clone);

        for(Node v : node.neighbors) {
            Node n = dfs(v, visited);
            clone.neighbors.add(n);
            // if(!visited.containsKey(v)) {
            //     //visited.offer(v);
            //     Node n = dfs(v, visited);
            //     clone.neighbors.add(n);
            // } else {
            //     clone.neighbors.add(visited.get(v));
            // }
        }
        return clone;
    }
}